stackrox/stackrox
# fix(scanner): Fix indexer auth when fetching layers

## Description

1. Setup transport with auth if provided in the `CreateIndexReport` request.
2. Add parameter to `scannerctl` to specify basic auth, for testing.
3. Opportunistic change: Change indexer implementation to local indexer, foreseeing we will have remote indexers for the Matcher later on.

| master
| .... :arrow_double_up: ROX-16402: Add converters for Matcher types #7823
| ........ :arrow_double_up: **fix(scanner): Fix indexer auth when fetching layers #7842** :point_left: 
| ............ :arrow_double_up: ROX-16402: Introduce the initial Scanner Matcher #7843

## Testing Performed

1. Verify auth is required.
2. Make calls with auth enabled.

### Verify authorization is required

Run scanner locally:

```
% cat scanner-config.yaml
indexer:
  db_conn_string: postgresql:///jvdm?host=/var/run/postgresql
matcher:
  db_conn_string: postgresql:///jvdm?host=/var/run/postgresql
mtls:
  certs_dir: /home/jvdm/tmp/scanner-tls
log_level: debug
%
% make -C scanner scanner-go-build
make: Entering directory '/home/jvdm/wt/github.com/stackrox/stackrox/scannerv4-client/scanner'
+ scanner-go-build
CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go build -trimpath -ldflags="-X github.com/stackrox/stackrox/scanner/internal/version.Version=4.2.x-185-gdcc0cb7747-dirty" -o image/scanner/bin/scanner ./cmd/scanner
make: Leaving directory '/home/jvdm/wt/github.com/stackrox/stackrox/scannerv4-client/scanner'
%
% scanner/image/scanner/bin/scanner -conf scanner-config.yaml
...
```

Run `scannerctl`, pass an image used for QA testing that requires authorization:

```
% go run scanner/cmd/scannerctl/main.go -certs ~/tmp/scanner-tls 'https://quay.io/rhacs-eng/qa:ose-jenkins'
2023/09/20 12:50:11 Reply: <nil> (rpc error: code = Internal desc = listing image layers (reference "quay.io/rhacs-eng/qa:ose-jenkins"): GET https://quay.io/v2/rhacs-eng/qa/manifests/ose-jenkins: UNAUTHORIZED: access to the requested resource is not authorized; map[])
%
```

Observe logs in `scanner`:

```
{"level":"error","host":"descending","component":"scanner/service/indexer","resource_type":"containerimage","manifest_digest":"sha512:1eb93f975af63c42633473db78f255ecb324aa0cb52a59440e8ee1aad4c08077077d5f28fc3a796c1af3ca453189eee6048c17acccd0c9bcb9ccbc40c40099e6","error":"listing image layers (reference \"quay.io/rhacs-eng/qa:ose-jenkins\"): GET https://quay.io/v2/rhacs-eng/qa/manifests/ose-jenkins: UNAUTHORIZED: access to the requested resource is not authorized; map[]","time":"2023-09-20T12:50:11-07:00"}
```

### Make calls with auth enabled

```
go run scanner/cmd/scannerctl/main.go -certs ~/tmp/scanner-tls -auth "$(cat ${XDG_RUNTIME_DIR}/containers/auth.json| jq -r '.auths["quay.io"].auth' | base64 -d)" 'https://quay.io/rhacs-eng/qa:ose-jenkins'
```

Observe output with the index report.

Expected scanner logs:

```
{"level":"info","host":"descending","component":"libindex/Libindex.Index","manifest":"sha512:1eb93f975af63c42633473db78f255ecb324aa0cb52a59440e8ee1aad4c08077077d5f28fc3a796c1af3ca453189eee6048c17acccd0c9bcb9ccbc40c40099e6","resource_type":"containerimage","manifest_digest":"sha512:1eb93f975af63c42633473db78f255ecb324aa0cb52a59440e8ee1aad4c08077077d5f28fc3a796c1af3ca453189eee6048c17acccd0c9bcb9ccbc40c40099e6","time":"2023-09-20T12:58:51-07:00","message":"index request done"}
```
